src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -156,11 +191,94 @@ public override void VisitMarkupTagHelperAttribute(MarkupTagHelperAttributeSynta
                 if (node.TagHelperAttributeInfo.Bound)
                 {
-                    _syntaxNodes.Add(node.Name);
+                    var result = new SyntaxResult(node.Name, SyntaxKind.MarkupTagHelperAttribute, _razorCodeDocument);
+                    _syntaxNodes.Add(result);
                 }
 
                 base.VisitMarkupTagHelperAttribute(node);
             }
+
+            public override void VisitMarkupTagHelperDirectiveAttribute(MarkupTagHelperDirectiveAttributeSyntax node)
+            {
+                if (node.TagHelperAttributeInfo.Bound)
+                {
+                    var transition = new SyntaxResult(node.Transition, SyntaxKind.Transition, _razorCodeDocument);
+                    _syntaxNodes.Add(transition);
+
+                    var directiveAttribute = new SyntaxResult(node.Name, SyntaxKind.MarkupTagHelperDirectiveAttribute, _razorCodeDocument);
+                    _syntaxNodes.Add(directiveAttribute);
+
+                    if (node.Colon != null)
+                    {
+                        var colon = new SyntaxResult(node.Colon, SyntaxKind.Colon, _razorCodeDocument);
+                        _syntaxNodes.Add(colon);
+                    }
+
+                    if (node.ParameterName != null)
+                    {
+                        var parameterName = new SyntaxResult(node.ParameterName, SyntaxKind.MarkupTagHelperDirectiveAttribute, _razorCodeDocument);
+                        _syntaxNodes.Add(parameterName);
+                    }
+                }
+
+                base.VisitMarkupTagHelperDirectiveAttribute(node);
+            }
+
+            public override void VisitMarkupMinimizedTagHelperDirectiveAttribute(MarkupMinimizedTagHelperDirectiveAttributeSyntax node)
+            {
+                if (node.TagHelperAttributeInfo.Bound)
+                {
+                    var transition = new SyntaxResult(node.Transition, SyntaxKind.Transition, _razorCodeDocument);
+                    _syntaxNodes.Add(transition);
+
+                    var directiveAttribute = new SyntaxResult(node.Name, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute, _razorCodeDocument);
+                    _syntaxNodes.Add(directiveAttribute);
+
+                    if (node.Colon != null)
+                    {
+                        var colon = new SyntaxResult(node.Colon, SyntaxKind.Colon, _razorCodeDocument);
+                        _syntaxNodes.Add(colon);
+                    }
+
+                    if (node.ParameterName != null)
+                    {
+                        var parameterName = new SyntaxResult(node.ParameterName, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute, _razorCodeDocument);
+                        _syntaxNodes.Add(parameterName);
+                    }
+                }
+
+                base.VisitMarkupMinimizedTagHelperDirectiveAttribute(node);
+            }
+
+            private bool ClassifyTagName(MarkupTagHelperElementSyntax node)
```

Comment:
This could just take in tagName and TagHelperInfo to avoid unnecessary casting.

Reply:
Actually we'll have to cast regardless to get the `TagHelperInfo` because that's only available on certain Syntax nodes (ie: It's on `MarkuptagHelperElementSyntax` but not on `MarkupTagHelperStartTagSyntax`). Given that (and that `MarkupTagHelperStartTagSyntax` and `MarkupTagHelperEndTagSyntax` don't share a (useful) common ancestor) my preference is still to pass a `MarkupTagHelperElementSyntax` here.

Reply:
Agree. From an offline conversation with @ryanbrandenburg, it looks like https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs#L29 might be used in place of this method but I'm not sure. @NTaylorMullen what do you think?

Reply:
Oh geez, I didn't even know we had that. I'm fine with giving that a shot to see if it covers all of the existing cases.

Reply:
I tried it, it #DidNotWork, BUT it did make me realize that I wasn't testing the scenario of 
```html
<p @onclick='Something'>
```
so I'm adding that test.